homeassistant/components/mqtt/client.py

```diff
@@ -895,10 +895,17 @@ def _async_mqtt_on_connect(
         import paho.mqtt.client as mqtt
 
         if result_code != mqtt.CONNACK_ACCEPTED:
+            if result_code in (
+                mqtt.CONNACK_REFUSED_BAD_USERNAME_PASSWORD,
+                mqtt.CONNACK_REFUSED_NOT_AUTHORIZED,
+            ):
+                self.hass.async_create_task(self.async_disconnect())
```

Comment:
```suggestion
                self._should_reconnect = False
                self.hass.async_create_task(self.async_disconnect())
```

Reply:
If we don't set this before calling `_async_connection_result` it will start trying to reconnect.

`async_disconnect` will do it but there is a chance it could suspend so it isn't set in time before we reach `_async_connection_result`